### PR TITLE
admission controllers: untangle global auto-instrumentation webhook functionality

### PR DESCRIPTION
**Pull request by stanistan**
Closes #26632. 

### What does this PR do?

This PR adds a couple of abstractions in order for other (config and tagsfromlabels) webhooks to not have an implicit dependency on autoinstrumentation. This makes the dependency explicit and puts it behind an interface.

Concretely:

- The `autoinstrumentation.ShouldInject` function is removed, its functionality is split into a few different functions and types, we are using `common.InjectionFilter.ShouldMutatePod(pod)` instead. 
  
  The type `common.InjectionFilter` uses a `NamespaceInjectionFilter` -- which was the logic that was baked into the `autoinstrumentation.Webhook` and has check `common IsExplicitPodMutationEnabled` (which was repeated multiple times throughout the code).

  This was all done so we can pass the `common.InjectionFilter` as an actual dependency to all of the webhooks that use it (`config`, `tagsfromlabels`, and `autoinstrumentation`) as a _parameter_.

- I kept the singleton pattern (which was the webhook) but moved it to _only_ be the namespace filtering functionality so that the webhook didn't need to be constructed in order for it to be used. This allowed us to only have a `NewWebhook` instead of that _and_ a `GetWebhook`.

- Because I kept the singleton I added test helpers for all of the places where we are using this implicit dependency to make it explicit in testing, as well as wrapping it in a helper `WithResetInjectionFilter*` family of functions that do the set up and cleanup of the quasi-global.

- `mutatingWebhooks` (in the controller creation flow) now takes this as an explicit parameter.

### Motivation

Right now, the `autoinstrumentation.Webhook` is a global dependency for other webhooks because namespace filtering for pod mutation is baked into its configuration. This built in complexity seeps throughout the code base... 

#### Example: Implict Webhook Construction

https://github.com/DataDog/datadog-agent/blob/dd93535b02b494e9f7a07cfcef5e1a4406f34b20/pkg/clusteragent/admission/mutate/config/config.go#L175-L177

This `ShouldInject` function duplicates logic in other functions (see: #26632 for a riveting diagram). 

#### Example: Odd test patterns

https://github.com/DataDog/datadog-agent/blob/dd93535b02b494e9f7a07cfcef5e1a4406f34b20/pkg/clusteragent/admission/controllers/webhook/controller_v1_test.go#L962-L963

### Possible Drawbacks / Trade-offs

This is still not _perfect_ but it is an iterative improvement that will allow us to re-use the filtering functionality in a safer and more explicit way.

### Describe how to test/QA your changes

Added more tests for the namespace injection filter, and made sure that the all the tests pass for the current controllers and mutating webhooks. Added more tests in auto-instrumentation to validate `mutate_unlabelled` on vs off behavior.

### CODE REVIEW

**Review by davidor — APPROVED** [review]

File: pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	} else {
+		filterExcludeList = append(apmDisabledNamespacesWithPrefix, disabledByDefault...)
+	}
+
+	return containers.NewFilter(containers.GlobalFilter, apmEnabledNamespacesWithPrefix, filterExcludeList)
+}
+
+// GetInjectionFilter is an accessor for the autoInstrumentationFilter
+func GetInjectionFilter() mutatecommon.NamespaceInjectionFilter {
+	return autoInstrumentationFilter
```
Comment (jonbodner) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go:138]: If we instantiate here, we could get rid of the sync.Once. I don't see other direct references in non-test code.

**Review by jonbodner — CHANGES_REQUESTED** [review]
I think the global var can go away, but I might be wrong.

File: pkg/clusteragent/admission/controllers/webhook/controller_base.go [hunk 2]
```diff
@@ -79,21 +80,31 @@ type MutatingWebhook interface {
...
 // now, the only restriction is that the agent sidecar webhook needs to go after
 // the config one. The reason is that the volume mount for the APM socket added
 // by the config webhook doesn't always work on Fargate (one of the envs where
 // we use an agent sidecar), and the agent sidecar webhook needs to remove it.
-func mutatingWebhooks(wmeta workloadmeta.Component, pa workload.PodPatcher) []MutatingWebhook {
+//
+// Note: the auto_instrumentation pod injection filter is used across
+// multiple mutating webhooks, so we add it as a hard dependency to each
+// of the components that use it via the injectionFilter parameter.
+func mutatingWebhooks(wmeta workloadmeta.Component, pa workload.PodPatcher, injectionFilter common.InjectionFilter) []MutatingWebhook {
```
[the remark this answers is not in the document]
Comment (stanistan) [reply pkg/clusteragent/admission/controllers/webhook/controller_base.go:99]: Right now it's not but I feel like it's nice to have the dependency to be as explicit as possible, and if we want to add more tests later to different filtering tests, it's much easier for all of the webhooks that use this.

File: pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	// The "config" and the "tags" webhooks depend on the "auto_instrumentation"
+	// configuration to decide if a pod should be injected.
+	//
+	// They first check if the pod has the label to enable mutations.
+	// If it doesn't, they mutate the pod if the option to "mutate_unlabeled" is
+	// set to true or if APM SSI is enabled in the namespace.
+	autoInstrumentationFilter = &injectionFilter{}
+)
+
+var _ mutatecommon.NamespaceInjectionFilter = &injectionFilter{}
```
[the remark this answers is not in the document]
Comment (stanistan) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go:34]: Yeap, will remove, unnecessary as tests and everything else will cover this.

File: pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go (hunk 1 above)
Comment (stanistan) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go:138]: I think you're right! 

I did most of the work to get rid of the synchronized global access but didn't finish the work. Passing the thing as an explicit dependency in webhook construction fixes most of this.

File: pkg/clusteragent/admission/controllers/webhook/controller_base.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stanistan) [reply pkg/clusteragent/admission/controllers/webhook/controller_base.go]: Actually making this change :)

File: pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	// We need a global variable to store the filter instance
+	// because other webhooks depend on it.
+	//
+	// The "config" and the "tags" webhooks depend on the "auto_instrumentation"
+	// configuration to decide if a pod should be injected.
+	//
+	// They first check if the pod has the label to enable mutations.
+	// If it doesn't, they mutate the pod if the option to "mutate_unlabeled" is
+	// set to true or if APM SSI is enabled in the namespace.
+	autoInstrumentationFilter = &injectionFilter{}
```
[the remark this answers is not in the document]
Comment (stanistan) [reply pkg/clusteragent/admission/mutate/autoinstrumentation/injection_filter.go]: Also removing the global.

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 5]
```diff
@@ -177,103 +164,37 @@ type Webhook struct {
...
-func NewWebhook(wmeta workloadmeta.Component) (*Webhook, error) {
-	filter, err := apmSSINamespaceFilter()
-	if err != nil {
+// NewWebhook returns a new Webhook dependent on the injection filter.
+func NewWebhook(wmeta workloadmeta.Component, filter mutatecommon.InjectionFilter) (*Webhook, error) {
+	// Note: the webhook is not functional with the filter being disabled--
+	//       and the filter is _global_! so we need to make sure that it was
+	//       initialized as it validates the configuration itself.
+	if filter.NSFilter == nil {
+		return nil, errors.New("filter required for auto_instrumentation webhook")
```
Comment (jonbodner) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:179]: we should probably use a sentinel error for this.

**Review by jonbodner — APPROVED** [review]
minor request, can ignore.

**Pull request merged**